File: source/Python/omegatomo/io/loadSkyscanData.py
# -*- coding: utf-8 -*-

import numpy as np
import os
import tkinter as tk
from tkinter.filedialog import askopenfilename
from omegatomo.io.loadProjectionImages import loadProjectionImages

def loadSkyscanData(options):
    def select_file():
    
        if len(options.fpath) == 0:
            root = tk.Tk()
            root.withdraw()
            nimi = askopenfilename(title='Select Skyscan log file',filetypes=([('log Files','*.log')]))
            if not nimi:
                raise ValueError('No file was selected')
        else:
            if not os.path.exists(options.fpath):
                print('Specified file was not found! Please select Nikon xtekct file')
                root = tk.Tk()
                root.withdraw()
                nimi = askopenfilename(title='Select Skyscan log file',filetypes=([('log Files','*.log')]))
                if not nimi:
                    raise ValueError('No file was selected')
            else:
                nimi = options.fpath
        
        file = os.path.split(nimi)[1]
        file, ext = os.path.splitext(file)
        fpath = os.path.split(nimi)[0] + '/'
        return file, fpath, nimi
    
    file, fpath, nimi = select_file()
    
    
    with open(nimi, 'r') as fid:
        hdr = [line.strip() for line in fid.readlines()]

    if not hasattr(options, 'binning'):
        options.binning = 1
    if not hasattr(options, 'only_reconstructions'):
        options.only_reconstructions = False


    for i in range(len(hdr)):
        if hdr[i].split("=")[0] == 'Number of Columns':
            options.nRowsD = int(hdr[i].split("=")[1]) // options.binning
        if hdr[i].split("=")[0] == 'Number of Rows':
            options.nColsD = int(hdr[i].split("=")[1]) // options.binning
        if hdr[i].split("=")[0] == 'Camera Pixel Size (um)':
            options.dPitchX = float(hdr[i].split("=")[1]) * options.binning / 1000
        if hdr[i].split("=")[0] == 'Camera Pixel Size (um)':
            options.dPitchY = float(hdr[i].split("=")[1]) * options.binning / 1000
        if hdr[i].split("=")[0] == 'Camera to Source (mm)':
            options.sourceToDetector = float(hdr[i].split("=")[1])
        if hdr[i].split("=")[0] == 'Object to Source (mm)':
            options.sourceToCRot = float(hdr[i].split("=")[1])
        if hdr[i].split("=")[0] == 'Number of Files':
            options.nProjections = int(hdr[i].split("=")[1]) - 1

    if not hasattr(options, 'only_reconstructions') or not options.only_reconstructions:
        file_path = os.path.join(fpath, file + '0000.tif')
        options.SinM = loadProjectionImages(options.nProjections, options.binning, file_path)
        options.SinM = np.transpose(options.SinM, (1, 0, 2))

File: source/m-files/loadSkyscanData.m
function options = loadSkyscanData(options)
%UNTITLED Summary of this function goes here
%   Detailed explanation goes here
if ~isfield(options, 'fpath') || isempty(options.fpath)
    [file, fpath] = uigetfile('*.log','Select Skyscan log file');
    if isequal(file, 0)
        error('No file was selected')
    end
    nimi = [fpath file];
    [fpath,file,ext] = fileparts(nimi);
else
    if exist(options.fpath, 'file') == 0
        warning(['Input file ' options.fpath ' not found!'])
        [file, fpath] = uigetfile('*.log','Select Skyscan log file');
        if isequal(file, 0)
            error('No file was selected')
        end
        nimi = [fpath file];
        [fpath,file,ext] = fileparts(nimi);
    else
        [fpath,file,ext] = fileparts(options.fpath);
        nimi = [fpath filesep file ext];
    end
    % nimi = options.fpath;
    % fpath = nimi(1:end-7);
end
fid = fopen(nimi);
hdr = textscan(fid,'%s','Delimiter','=');
hdr = hdr{1};
fclose(fid);
if isfield(options, 'only_reconstructions') 
    only_reconstructions = options.only_reconstructions;
else
    only_reconstructions = false;
end
ind = find(strcmp(hdr,'Number of Columns')) + 1;
options.nRowsD = str2double(hdr{ind(1)}) / options.binning;
ind = find(strcmp(hdr,'Number of Rows')) + 1;
options.nColsD = str2double(hdr{ind(1)}) / options.binning;
ind = find(strcmp(hdr,'Camera Pixel Size (um)')) + 1;
options.dPitchX = str2double(hdr{ind(1)}) * options.binning / 1000;
ind = find(strcmp(hdr,'Camera Pixel Size (um)')) + 1;
options.dPitchY = str2double(hdr{ind(1)}) * options.binning / 1000;
ind = find(strcmp(hdr,'Camera to Source (mm)')) + 1;
options.sourceToDetector = str2double(hdr{ind(1)});
ind = find(strcmp(hdr,'Object to Source (mm)')) + 1;
options.sourceToCRot = str2double(hdr{ind(1)});
ind = find(strcmp(hdr,'Number of Files')) + 1;
options.nProjections = str2double(hdr{ind(1)}) - 1;
if ~only_reconstructions || ~isfield(options,'SinM')
    fpath = [fpath filesep file '0000.tif'];
    options.SinM = loadProjectionImages(options.nProjections,options.binning, fpath);
    options.SinM = permute(options.SinM, [2 1 3]);
end
end